### ISSUE TITLE
tilesets are not always listed in the tmx map file.

### ISSUE THREAD

**Issue by AdrienLemaire** [mapeditor/tiled#3017]
[Related discussion](https://discord.com/channels/524610627545595904/524610627545595906/827360977740627998) with @eishiya

On a multi-map project, when adding an object with a template referencing a spritesheet tileset, that tileset was missing in the saved .tmx file.

tmx file
```
<map version="1.5" tiledversion="1.5.0" orientation="orthogonal" renderorder="right-down" width="65" height="65" tilewidth="128" tileheight="128" infinite="0" nextlayerid="111" nextobjectid="1202">
 <tileset firstgid="1" source="tiles/forest-terrain.tsx"/>
 <tileset firstgid="1025" source="objects/forest-trees.tsx"/>
 <tileset firstgid="1043" source="objects/basic-objects.tsx"/>
 <tileset firstgid="1054" source="tiles/animations/grass-var-04.tsx"/>
 <tileset firstgid="1070" source="tiles/animations/grass-var-03.tsx"/>
 <tileset firstgid="1086" source="objects/village-items.tsx"/>
 ...
 <objectgroup id="109" name="Player" offsetx="512" offsety="-256">
  <object id="1157" template="templates/player-character.tx" x="6331.43" y="3680.57"/>
 </objectgroup>
</map>
```
player-character.tx
```
<?xml version="1.0" encoding="UTF-8"?>
<template>
 <tileset firstgid="1" source="../tiles/animations/player-animations-01.tsx"/>
 <object name="Player" type="Player" gid="1" width="128" height="128"/>
</template>
```

Adding manually ` <tileset firstgid="1102" source="tiles/animations/player-animations-01.tsx"/>`  in the tmx file solved our issues. We were expecting Tiled to always list used tilesets in the tmx file.

**Comment by bjorn** [comment]
Why do you need the tileset to be referenced by the map? How do you make sure the template instance is using the right tileset?

Consider the case when the tileset would be listed in the map. It could not be used for resolving the tile reference of the template instance, because the tileset reference in the map file might have a different `firstgid` than the one in the template file. There would be no way to know which tileset the template instance is referencing without reading the template file. And if we need to read the template file anyway, there is no point in listing this tileset in the map file as well.

From the discussion on Discord I noticed that you've tried using the "Detach templates" feature. In this case, the tileset _should_ get listed in the map file, and the fact that it doesn't is a bug.

**Comment by bjorn** [comment]
> From the discussion on Discord I noticed that you've tried using the "Detach templates" feature. In this case, the tileset _should_ get listed in the map file, and the fact that it doesn't is a bug.

This bug was fixed in 310f79dc197c65b8df9b395864afffec98e7d92d.
